### ISSUE TITLE
Questions regarding TERA

### ISSUE THREAD

**Issue by JunwenBai** [andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning#27]
I have several questions regarding your TERA paper:
1. You used masked reconstruction for 80%, noisy reconstruction for 10% and clean reconstruction for 10%. Is this strategy better than 100% masked reconstruction? By how much?
2. Your feature extractor is a transformer and you added liGRU on top of it for finetuning. Why you used GRU instead of transformers for finetuning? Does GRU perform better?
3. You also mentioned in the paper that not freezing TERA in finetuning performs better. By not freezing, do you mean you don't freeze it from the beginning, or you freeze TERA for a while and then defreeze it in the later epochs?
4. To understand your TERA code, which files should I look into? It seems that they are wrapped quite deep?
Thx!

**Comment by andi611** [comment]
Hi,

> I have several questions regarding your TERA paper:
> 
> 1. You used masked reconstruction for 80%, noisy reconstruction for 10% and clean reconstruction for 10%. Is this strategy better than 100% masked reconstruction? By how much?

The 80% / 10% / 10% masking strategies along the time axis is from [BERT](https://arxiv.org/abs/1810.04805), it is better than 100% according to the BERT paper. This also makes sense, as the 10% clean allows the model to see clean data during training.

> 2. Your feature extractor is a transformer and you added liGRU on top of it for finetuning. Why you used GRU instead of transformers for finetuning? Does GRU perform better?

We also explore adding a single MLP layer (which is basically fine-tuning just the transformers). In our exp we find liGRU to perform better than MLP, at the cost of significantly longer training time (due to liGRU's recurrence).

> 3. You also mentioned in the paper that not freezing TERA in finetuning performs better. By not freezing, do you mean you don't freeze it from the beginning, or you freeze TERA for a while and then defreeze it in the later epochs?

We don't freeze from the beginning, we did not explore defreezing strategies.

> 4. To understand your TERA code, which files should I look into? It seems that they are wrapped quite deep?

The code related to input alterations (time + freq + mag) can be found in [transformer/mam.py](https://github.com/andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning/blob/f25379b77a9e9bd5c6278ebc625bb60022a17cf0/transformer/mam.py#L80-L188), and the code of model architecture can be found in [transformer/model.py](https://github.com/andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning/blob/f25379b77a9e9bd5c6278ebc625bb60022a17cf0/transformer/model.py#L396-L582)
>    Thx!

I hope this helps!

**Comment by JunwenBai** [comment]
Thx for the reply! Besides, for the alternation strategy, why did you ignored C in Fig. 2?

**Comment by andi611** [comment]
> Thx for the reply!
> Besides, for the alternation strategy, why did you ignored C in Fig. 2? 

It is **NOT** ignored, it is incorporated in the time alteration objective (80% mask / **10% replace** / 10% clean). Fig. 2C is the 10% replacment. Implemented [here](https://github.com/andi611/Self-Supervised-Speech-Pretraining-and-Representation-Learning/blob/f25379b77a9e9bd5c6278ebc625bb60022a17cf0/transformer/mam.py#L151-L155).

> And, how long does it typically take for pretrain+finetuning with your code?

It depends on your GPU and the amount of pre-train data. In my case, 100 hrs of pre-train data takes a day to pre-train, 460 hrs take around 3 days, 960 hrs take around 5 days. Finetuning with liGRU takes a week, with MLP it takes only a couple hours (less than a day).

**Comment by JunwenBai** [comment]
This is with one 1080Ti GPU?

**Comment by andi611** [comment]
> This is with one 1080Ti GPU?

Yes!

**Comment by JunwenBai** [comment]
One more question: when you evaluate WER, do you consider 'color' and 'colour' are the same? Namely, the American English and British English?

**Comment by andi611** [comment]
> When you evaluate WER, do you consider 'color' and 'colour' are the same? Namely, the American English and British English?

Since all the label data are from LibriSpeech, hence no we do not consider the difference between American English and British English. (FYI, we use the standard Kaldi pipeline for evaluating WER)
